File: tests/csharp_static/TestValueCarriers.cs
using System.Text.Json;
using System.Text.Json.Nodes;
using MyAirtable;
using Xunit;

namespace MyAirtable.Tests;

/// <summary>
/// CS5.1/5.2/5.4 — value-carrier (de)serialization. The carriers carry no [JsonPropertyName],
/// so the global camelCase naming policy maps them to Airtable's wire shape (url, profilePicUrl,
/// …). Critical for writes: a URL-based attachment create sends <c>{"url": ...}</c> and a
/// collaborator write sends <c>{"id": ...}</c>.
/// </summary>
public class TestValueCarriers
{
    private static string Encode<T>(T value) =>
        JsonSerializer.Serialize(value, AirtableJson.Options);

    private static T Decode<T>(string json) =>
        JsonSerializer.Deserialize<T>(json, AirtableJson.Options)!;

    // ---- attachments (CS5.1) ----

    [Fact]
    public void AttachmentUrlOnlyEncodesToCamelCaseUrlOnly()
    {
        var node = JsonNode.Parse(
            Encode(new AirtableAttachment { Url = "https://example.com/a.png" })
        )!;
        Assert.Equal("https://example.com/a.png", node["url"]!.GetValue<string>());
        // null fields omitted (WhenWritingNull) — a clean URL-based upload payload.
        Assert.Null(node["id"]);
        Assert.Null(node["filename"]);
    }

    [Fact]
    public void AttachmentDecodesFullWireShapeIncludingThumbnails()
    {
        var model = Decode<AirtableAttachment>(
            """
            {
              "id": "att123",
              "url": "https://example.com/a.png",
              "filename": "a.png",
              "size": 2048,
              "type": "image/png",
              "thumbnails": { "small": { "url": "https://example.com/s.png", "width": 36, "height": 36 } }
            }
            """
        );
        Assert.Equal("att123", model.Id);
        Assert.Equal("a.png", model.Filename);
        Assert.Equal(2048, model.Size);
        Assert.Equal("image/png", model.Type);
        Assert.Equal("https://example.com/s.png", model.Thumbnails!.Small!.Url);
        Assert.Equal(36, model.Thumbnails.Small.Width);
    }

    // ---- collaborators (CS5.2) ----

    [Fact]
    public void CollaboratorIdOnlyEncodesToCamelCaseIdOnly()
    {
        var node = JsonNode.Parse(Encode(new AirtableCollaborator { Id = "usr123" }))!;
        Assert.Equal("usr123", node["id"]!.GetValue<string>());
        Assert.Null(node["email"]);
        Assert.Null(node["name"]);
    }

    [Fact]
    public void CollaboratorDecodesProfilePicUrlFromCamelCase()
    {
        var model = Decode<AirtableCollaborator>(
            """
            { "id": "usr1", "email": "a@b.c", "name": "Ada", "profilePicUrl": "https://example.com/p.png" }
            """
        );
        Assert.Equal("usr1", model.Id);
        Assert.Equal("a@b.c", model.Email);
        Assert.Equal("Ada", model.Name);
        Assert.Equal("https://example.com/p.png", model.ProfilePicUrl);
    }

    // ---- buttons (CS5.4) ----

    [Fact]
    public void ButtonDecodesLabelAndUrl()
    {
        var model = Decode<AirtableButton>("""{ "label": "Open", "url": "https://example.com" }""");
        Assert.Equal("Open", model.Label);
        Assert.Equal("https://example.com", model.Url);
    }
}

File: tests/cpp_static/test_value_carriers.cpp
// Value-carrier (de)serialization (C# TestValueCarriers parity).
#include <catch2/catch_test_macros.hpp>

#include "airtable_attachment.hpp"
#include "airtable_button.hpp"
#include "airtable_collaborator.hpp"
#include "airtable_record.hpp"
#include "sort.hpp"

using namespace myairtable;

TEST_CASE("attachment decodes full metadata incl nested thumbnails", "[carriers]") {
    auto a = json::parse(R"({
        "id": "att123", "url": "https://x/y.png", "filename": "y.png",
        "size": 1024, "type": "image/png",
        "thumbnails": {"small": {"url": "https://x/s.png", "width": 36, "height": 36},
                        "large": {"url": "https://x/l.png", "width": 512, "height": 512}}
    })")
                 .get<AirtableAttachment>();
    REQUIRE(a.id == "att123");
    REQUIRE(a.url == "https://x/y.png");
    REQUIRE(a.filename == "y.png");
    REQUIRE(a.size == 1024);
    REQUIRE(a.type == "image/png");
    REQUIRE(a.thumbnails->small->width == 36);
    REQUIRE(a.thumbnails->large->url == "https://x/l.png");
    REQUIRE(a.thumbnails->full == std::nullopt);
}

TEST_CASE("attachment upload shape is url-only, no null keys", "[carriers]") {
    auto upload = AirtableAttachment{.url = "https://example.com/file.pdf"};
    json encoded = upload;
    REQUIRE(encoded == json::parse(R"({"url":"https://example.com/file.pdf"})"));
}

TEST_CASE("collaborator decodes and id-only write shape round-trips", "[carriers]") {
    auto c =
        json::parse(R"({"id":"usr1","email":"a@b.co","name":"Ann","profilePicUrl":"https://p"})")
            .get<AirtableCollaborator>();
    REQUIRE(c.id == "usr1");
    REQUIRE(c.email == "a@b.co");
    REQUIRE(c.name == "Ann");
    REQUIRE(c.profile_pic_url == "https://p");

    json write_shape = AirtableCollaborator{.id = "usr1"};
    REQUIRE(write_shape == json::parse(R"({"id":"usr1"})"));
}

TEST_CASE("button decodes label and url", "[carriers]") {
    auto b = json::parse(R"({"label":"Open","url":"https://go"})").get<AirtableButton>();
    REQUIRE(b.label == "Open");
    REQUIRE(b.url == "https://go");
    // Airtable omits url when the button formula yields none.
    auto no_url = json::parse(R"({"label":"Open"})").get<AirtableButton>();
    REQUIRE(no_url.url == std::nullopt);
}

TEST_CASE("record envelope decodes id, createdTime, raw fields", "[carriers]") {
    auto r = json::parse(R"({
        "id": "recABC", "createdTime": "2024-01-15T10:30:00.000Z",
        "fields": {"fldX": "hello", "fldY": 5}
    })")
                 .get<AirtableRecord>();
    REQUIRE(r.id == "recABC");
    REQUIRE(r.created_time.epoch_millis() == 1'705'314'600'000);
    REQUIRE(r.fields.at("fldX") == "hello");
    // Fields default to an empty object when the record carries none.
    REQUIRE(json::parse(R"({"id":"recE"})").get<AirtableRecord>().fields == json::object());
}

TEST_CASE("sort direction wire values", "[carriers]") {
    REQUIRE(wire(SortDirection::Asc) == "asc");
    REQUIRE(wire(SortDirection::Desc) == "desc");
    Sort s{.field = "Name"};
    REQUIRE(s.direction == SortDirection::Asc); // default ascending
}
